### DISCUSSION TITLE
Exponential parsing time with nested expressions

### DISCUSSION THREAD

**Discussion by ebastien** [zesterer/chumsky#489] [Q&A] [answered]
I am facing exponential parsing time for a language of mine when dealing with deeply nested expressions.
I suspect a pathological case in the way my grammar is written, leading to an unreasonable amount of lookahead/backtracking.
Reducing the test case, it seems to point to the way I express some of my postfix operators. I am not making use of the precedence climbing yet and expressing precedence by combining parsers explicitly (i.e. `expr = sum(product(term))`).
Before I start rewriting the grammar, I would like to know whether Chumsky implements any kind of memoization (packrat?) to maintain linear parsing time or if the optimization is left to the author of the grammar?

Thanks for the tremendous amount of work that went into this project btw!

PS: the parser code is [here](https://github.com/ebastien/openapi-lang/blob/master/oal-syntax/src/parser.rs#L638). Note that it relies on some abstractions of mine to retain all the tokens in a concrete syntax tree. That makes the grammar a bit more convoluted than I would like it to be. The pathological case is something like `let a = { 'p { 'p { 'p { 'p { 'p { 'p {} } } } } } };`.

**Comment by Zij-IT** [comment, marked as answer]
As you correctly identified, this is due to the how the grammar is structured, and chumsky repeatedly parsing the same piece again and again. I have three solutions for you that I can think of:

- Manually implement precedence climbing
  - On the current released version of Chumsky there isn't another method besides precedence climbing that you can use. Although this example is more complex, [this](https://github.com/zesterer/tao/blob/master/syntax/src/parse.rs#L628-L821) is how @zesterer used precedence climbing in his language Tao.
- Update to 1.0.4-alpha for `memoization`
  - On the alpha version of chumsky, there is an appropriately named feature `memoization`, though I personally have no experience with it, and @zesterer would probably be able to give you more pointers here.
- Update to the master branch of chumsky for `pratt`
  - If you are willing to be even more adventurous, on the master branch for chumsky there is a feature called `pratt` which enables you use the `pratt` method, and not have to deal with implementing precedence climbing yourself. This is probably where I can help out the most, as I wrote the implementation. Kinda looks like this:

```rust
 use chumsky::prelude::*;
 use chumsky::pratt::{left_infix, right_infix};

 enum Expr {
     Literal(i64),
     Add(Box<Expr>, Box<Expr>),
     Sub(Box<Expr>, Box<Expr>),
     Mul(Box<Expr>, Box<Expr>),
     Div(Box<Expr>, Box<Expr>),
 }

 impl std::fmt::Display for Expr {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
             Self::Literal(literal) => write!(f, "{literal}"),
             Self::Add(left, right) => write!(f, "({left} + {right})"),
             Self::Sub(left, right) => write!(f, "({left} - {right})"),
             Self::Mul(left, right) => write!(f, "({left} * {right})"),
             Self::Div(left, right) => write!(f, "({left} / {right})"),
         }
     }
 }

 let atom = text::int::<_, _, extra::Default>(10)
     .from_str()
     .unwrapped()
     .map(Expr::Literal);

 let operator = choice((
     left_infix(just('+'), 0, |l, r| Expr::Add(Box::new(l), Box::new(r))),
     left_infix(just('-'), 0, |l, r| Expr::Sub(Box::new(l), Box::new(r))),
     right_infix(just('*'), 1, |l, r| Expr::Mul(Box::new(l), Box::new(r))),
     right_infix(just('/'), 1, |l, r| Expr::Div(Box::new(l), Box::new(r))),
));

let expr = atom.pratt(operator.padded_by(just(' ')));
let expr_str = expr.map(|expr| expr.to_string()).then_ignore(end());
assert_eq!(expr_str.parse("1 + 2").into_result(), Ok("(1 + 2)".to_string()));
// `*` binds more strongly than `+`
assert_eq!(expr_str.parse("1 * 2 + 3").into_result(), Ok("((1 * 2) + 3)".to_string()));
assert_eq!(expr_str.parse("1 + 2 * 3").into_result(), Ok("(1 + (2 * 3))".to_string()));
// `+` is left-associative
assert_eq!(expr_str.parse("1 + 2 + 3").into_result(), Ok("((1 + 2) + 3)".to_string()));
// `*` is right-associative (in this example)
assert_eq!(expr_str.parse("1 * 2 * 3").into_result(), Ok("(1 * (2 * 3))".to_string()));
```

Moving up to the latest branch isn't as unstable as it sounds, as we are hoping to get a release out "soon" so that the benefits of zero copy parsing are available to the public. I have personally been using the branch for the last 4+ months and haven't had any troubles thus far. You could also pull a copy locally of the master branch, and just stay attached to whatever that version is.

TL;DR: Your options are update or implement precedence climbing manually

If anything in this wall is unclear, please let me know so I can get it all cleared up
